datafusion/common/src/tree_node.rs

```diff
@@ -551,22 +551,48 @@ impl<I: Iterator> TransformedIterator for I {
     ) -> Result<Transformed<Vec<Self::Item>>> {
         let mut tnr = TreeNodeRecursion::Continue;
         let mut transformed = false;
-        let data = self
-            .map(|item| match tnr {
-                TreeNodeRecursion::Continue | TreeNodeRecursion::Jump => {
-                    f(item).map(|result| {
-                        tnr = result.tnr;
-                        transformed |= result.transformed;
-                        result.data
-                    })
-                }
-                TreeNodeRecursion::Stop => Ok(item),
-            })
-            .collect::<Result<Vec<_>>>()?;
-        Ok(Transformed::new(data, transformed, tnr))
+        self.map(|item| match tnr {
+            TreeNodeRecursion::Continue | TreeNodeRecursion::Jump => {
+                f(item).map(|result| {
+                    tnr = result.tnr;
+                    transformed |= result.transformed;
+                    result.data
+                })
+            }
+            TreeNodeRecursion::Stop => Ok(item),
+        })
+        .collect::<Result<Vec<_>>>()
+        .map(|data| Transformed::new(data, transformed, tnr))
     }
 }
 
+/// Transformation helper to process sequence of tree node containing expressions.
+/// This macro is very similar to [TransformedIterator::map_until_stop_and_collect] to
+/// process nodes that are siblings, but it accepts an initial transformation and a
+/// sequence of pairs of an expression and its transformation.
```

Comment:
This is quite clever. I have some suggestions on naming that would have helped me:

Use `F` to mirror the nomenclature of `TransformedIterator`. I thought `$TRANSFORMED_EXPR`  was an actually expr (but it is actually a closure that is invoked I think 🤔 )

```
$TRANSFORMED_EXPR_0 --> F0      
$TRANSFORMED_EXPR --> F.
```

I think it would help to document each parameter and the return value (specifically it looks like it returns `Transformed<(data, ..., data)>`?

Reply:
Sure, let me fix these and come back yo you.

Reply:
Renamed in https://github.com/apache/arrow-datafusion/pull/9876/commits/a36f1aa02dd0652d9e514ba865a1f58db0d4f233